File: Swifty/Services/TripRecorder.swift
import Foundation
import CoreLocation
import Observation

/// 측정 시작부터 종료까지 경로와 통계를 모은다.
///
/// LocationService의 누적값을 빼서 쓰지 않고 직접 모으는 이유는,
/// 사용자가 측정 도중에 주행 기록을 초기화해도 측정이 깨지지 않게 하기 위해서다.
@MainActor
@Observable
final class TripRecorder {

    private(set) var isRecording = false
    private(set) var startedAt: Date?
    private(set) var distance: CLLocationDistance = 0
    private(set) var maxSpeed: Double = 0
    private(set) var elevationGain: Double = 0
    private(set) var elevationLoss: Double = 0
    private(set) var sampleCount = 0

    /// 측정을 시작한 이동수단. 도중에 바꿔도 기록에는 시작 시점 값을 남긴다.
    private(set) var transport: TransportMode = .car

    private var samples: [TripSample] = []
    private var distanceAnchor: CLLocation?
    private var altitudeAnchor: Double?
    private var lastSampleAt: Date?
    private var lastSampleLocation: CLLocation?

    /// 경로를 너무 촘촘히 남기면 파일만 커진다. 이 두 조건 중 하나는 넘겨야 한 점을 남긴다.
    private let minSampleInterval: TimeInterval = 2
    private let minSampleDistance: CLLocationDistance = 5
    private let usableVerticalAccuracy: CLLocationAccuracy = 15

    /// 측정 중 경과 시간. 화면 갱신은 LocationService의 1초 시계가 이끈다.
    func elapsed(now: Date) -> TimeInterval {
        guard let startedAt else { return 0 }
        return now.timeIntervalSince(startedAt)
    }

    var averageSpeed: Double {
        guard let startedAt else { return 0 }
        let duration = Date().timeIntervalSince(startedAt)
        guard duration > 1, distance > 0 else { return 0 }
        return distance / duration
    }

    // MARK: - 제어

    func start(transport: TransportMode, from location: CLLocation?) {
        guard !isRecording else { return }
        isRecording = true
        startedAt = .now
        self.transport = transport

        distance = 0
        maxSpeed = 0
        elevationGain = 0
        elevationLoss = 0
        sampleCount = 0
        samples = []
        distanceAnchor = location
        altitudeAnchor = nil
        lastSampleAt = nil
        lastSampleLocation = nil

        if let location { append(location, force: true) }
    }

    /// 측정을 끝내고 기록을 만든다. 점이 거의 없으면 저장할 게 없으므로 nil.
    func stop(destinationName: String?) -> TripRecord? {
        guard isRecording, let startedAt else { return nil }
        isRecording = false
        self.startedAt = nil

        defer { samples = [] }
        guard samples.count >= 2 else { return nil }

        return TripRecord(id: UUID(),
                          startedAt: startedAt,
                          endedAt: .now,
                          transport: transport,
                          distance: distance,
                          maxSpeed: maxSpeed,
                          elevationGain: elevationGain,
                          elevationLoss: elevationLoss,
                          destinationName: destinationName,
                          samples: samples)
    }

    func cancel() {
        isRecording = false
        startedAt = nil
        samples = []
    }

    // MARK: - 수집

    /// 새 위치가 들어올 때마다 호출한다.
    func ingest(_ location: CLLocation, smoothedSpeed: Double) {
        guard isRecording else { return }

        accumulateDistance(to: location)
        accumulateElevation(from: location)
        maxSpeed = max(maxSpeed, max(location.speed, 0))
        append(location, force: false, speed: smoothedSpeed)
    }

    private func append(_ location: CLLocation, force: Bool, speed: Double? = nil) {
        guard let startedAt else { return }

        if !force {
            let farEnough = lastSampleLocation.map {
                location.distance(from: $0) >= minSampleDistance
            } ?? true
            let longEnough = lastSampleAt.map {
                location.timestamp.timeIntervalSince($0) >= minSampleInterval
            } ?? true
            guard farEnough || longEnough else { return }
        }

        let altitude = (location.verticalAccuracy >= 0
                        && location.verticalAccuracy <= usableVerticalAccuracy)
            ? location.altitude : nil

        samples.append(TripSample(t: location.timestamp.timeIntervalSince(startedAt),
                                  lat: location.coordinate.latitude,
                                  lon: location.coordinate.longitude,
                                  speed: speed ?? max(location.speed, 0),
                                  altitude: altitude))
        sampleCount = samples.count
        lastSampleAt = location.timestamp
        lastSampleLocation = location
    }

    /// 앵커 방식. 잡음 수준을 넘어설 만큼 멀어졌을 때만 거리를 확정한다.
    private func accumulateDistance(to location: CLLocation) {
        guard let anchor = distanceAnchor else {
            distanceAnchor = location
            return
        }
        let delta = location.distance(from: anchor)
        let elapsed = location.timestamp.timeIntervalSince(anchor.timestamp)
        // 물리적으로 불가능한 순간 이동은 버리고 앵커만 옮긴다.
        if elapsed > 0, delta / elapsed > 111 {
            distanceAnchor = location
            return
        }
        let noise = max((anchor.horizontalAccuracy + location.horizontalAccuracy) / 2, 2)
        guard delta > noise else { return }
        distance += delta
        distanceAnchor = location
    }

    private func accumulateElevation(from location: CLLocation) {
        guard location.verticalAccuracy >= 0,
              location.verticalAccuracy <= usableVerticalAccuracy else { return }
        let altitude = location.altitude
        guard let anchor = altitudeAnchor else {
            altitudeAnchor = altitude
            return
        }
        let delta = altitude - anchor
        guard abs(delta) > max(location.verticalAccuracy, 3) else { return }
        if delta > 0 { elevationGain += delta } else { elevationLoss -= delta }
        altitudeAnchor = altitude
    }
}

File: android/app/src/main/java/com/gutaeho/swifty/data/TripRecorder.kt
package com.gutaeho.swifty.data

import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableIntStateOf
import androidx.compose.runtime.mutableDoubleStateOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import com.gutaeho.swifty.domain.Geo
import com.gutaeho.swifty.domain.LocationFix
import com.gutaeho.swifty.domain.TransportMode
import com.gutaeho.swifty.domain.TripRecord
import com.gutaeho.swifty.domain.TripSample
import kotlin.math.abs
import kotlin.math.max

class TripRecorder {
    var isRecording by mutableStateOf(false)
        private set
    var startedAtMillis by mutableStateOf<Long?>(null)
        private set
    var distanceMeters by mutableDoubleStateOf(0.0)
        private set
    var maxSpeedMetersPerSecond by mutableDoubleStateOf(0.0)
        private set
    var elevationGainMeters by mutableDoubleStateOf(0.0)
        private set
    var elevationLossMeters by mutableDoubleStateOf(0.0)
        private set
    var sampleCount by mutableIntStateOf(0)
        private set
    var transport by mutableStateOf(TransportMode.CAR)
        private set

    private var samples = mutableListOf<TripSample>()
    private var distanceAnchor: LocationFix? = null
    private var altitudeAnchor: Double? = null
    private var lastSampleAtMillis: Long? = null
    private var lastSampleLocation: LocationFix? = null

    val averageSpeedMetersPerSecond: Double
        get() {
            val duration = elapsedSeconds(System.currentTimeMillis())
            return if (duration > 1 && distanceMeters > 0) distanceMeters / duration else 0.0
        }

    fun elapsedSeconds(nowMillis: Long): Double = startedAtMillis?.let { (nowMillis - it).coerceAtLeast(0) / 1000.0 } ?: 0.0

    fun start(transport: TransportMode, location: LocationFix?) {
        if (isRecording) return
        isRecording = true
        startedAtMillis = System.currentTimeMillis()
        this.transport = transport
        distanceMeters = 0.0
        maxSpeedMetersPerSecond = 0.0
        elevationGainMeters = 0.0
        elevationLossMeters = 0.0
        samples.clear()
        sampleCount = 0
        distanceAnchor = location
        altitudeAnchor = null
        lastSampleAtMillis = null
        lastSampleLocation = null
        location?.let { append(it, true, it.speedMetersPerSecond) }
    }

    fun stop(destinationName: String?): TripRecord? {
        val started = startedAtMillis ?: return null
        if (!isRecording) return null
        isRecording = false
        startedAtMillis = null
        val completedSamples = samples.toList()
        samples.clear()
        if (completedSamples.size < 2) return null
        return TripRecord(
            startedAtMillis = started,
            endedAtMillis = System.currentTimeMillis(),
            transport = transport,
            distanceMeters = distanceMeters,
            maxSpeedMetersPerSecond = maxSpeedMetersPerSecond,
            elevationGainMeters = elevationGainMeters,
            elevationLossMeters = elevationLossMeters,
            destinationName = destinationName,
            samples = completedSamples,
        )
    }

    fun cancel() {
        isRecording = false
        startedAtMillis = null
        samples.clear()
        sampleCount = 0
    }

    fun ingest(location: LocationFix, smoothedSpeed: Double) {
        if (!isRecording) return
        accumulateDistance(location)
        accumulateElevation(location)
        maxSpeedMetersPerSecond = max(maxSpeedMetersPerSecond, max(location.speedMetersPerSecond, 0.0))
        append(location, false, smoothedSpeed)
    }

    private fun append(location: LocationFix, force: Boolean, speed: Double) {
        val started = startedAtMillis ?: return
        if (!force) {
            val farEnough = lastSampleLocation?.let { Geo.distance(it.coordinate, location.coordinate) >= 5 } ?: true
            val longEnough = lastSampleAtMillis?.let { location.timestampMillis - it >= 2_000 } ?: true
            if (!farEnough && !longEnough) return
        }
        samples += TripSample(
            elapsedSeconds = (location.timestampMillis - started) / 1000.0,
            coordinate = location.coordinate,
            speedMetersPerSecond = speed,
            altitudeMeters = location.altitudeMeters,
        )
        sampleCount = samples.size
        lastSampleAtMillis = location.timestampMillis
        lastSampleLocation = location
    }

    private fun accumulateDistance(location: LocationFix) {
        val anchor = distanceAnchor
        if (anchor == null) {
            distanceAnchor = location
            return
        }
        val delta = Geo.distance(anchor.coordinate, location.coordinate)
        val elapsed = (location.timestampMillis - anchor.timestampMillis) / 1000.0
        if (elapsed > 0 && delta / elapsed > 111) {
            distanceAnchor = location
            return
        }
        val noise = max((anchor.horizontalAccuracy + location.horizontalAccuracy) / 2, 2.0)
        if (delta > noise) {
            distanceMeters += delta
            distanceAnchor = location
        }
    }

    private fun accumulateElevation(location: LocationFix) {
        val altitude = location.altitudeMeters ?: return
        val accuracy = location.verticalAccuracy ?: return
        val anchor = altitudeAnchor
        if (anchor == null) {
            altitudeAnchor = altitude
            return
        }
        val delta = altitude - anchor
        if (abs(delta) <= max(accuracy, 3.0)) return
        if (delta > 0) elevationGainMeters += delta else elevationLossMeters -= delta
        altitudeAnchor = altitude
    }
}
